pkg/server/api/bundle/v1/service.go

```diff
@@ -65,17 +59,96 @@ func (s *Service) GetBundle(ctx context.Context, req *bundle.GetBundleRequest) (
 		return nil, status.Error(codes.NotFound, "bundle not found")
 	}
 
-	b, err := applyMask(dsResp.Bundle, req.OutputMask)
+	bundle, err := api.BundleToProto(dsResp.Bundle)
 	if err != nil {
-		log.Errorf("Failed to apply mask: %v", err)
-		return nil, status.Errorf(codes.Internal, "failed to apply mask: %v", err)
+		log.WithError(err).Error("Failed to convert bundle")
+		return nil, status.Errorf(codes.Internal, "failed to convert bundle: %v", err)
 	}
 
-	return b, nil
+	applyBundleMask(bundle, req.OutputMask)
+	return bundle, nil
 }
 
 func (s *Service) AppendBundle(ctx context.Context, req *bundle.AppendBundleRequest) (*types.Bundle, error) {
-	return nil, status.Errorf(codes.Unimplemented, "method AppendBundle not implemented")
+	log := rpccontext.Logger(ctx)
+
+	if req.Bundle == nil {
+		log.Error("Invalid request: missing bundle")
+		return nil, status.Error(codes.InvalidArgument, "missing bundle")
+	}
+
+	td, err := spiffeid.TrustDomainFromString(req.Bundle.TrustDomain)
+	if err != nil {
+		log.WithError(err).Errorf("Trust domain argument is not a valid SPIFFE ID: %q", req.Bundle.TrustDomain)
+		return nil, status.Errorf(codes.InvalidArgument, "trust domain argument is not a valid SPIFFE ID: %q", req.Bundle.TrustDomain)
+	}
+
+	if s.td.Compare(td) != 0 {
+		log.Errorf("Invalid request: Trust domain is not allowed: %q", req.Bundle.TrustDomain)
+		return nil, status.Errorf(codes.InvalidArgument, "trust domain is not allowed: %q", req.Bundle.TrustDomain)
+	}
+
+	dsResp, err := s.ds.FetchBundle(ctx, &datastore.FetchBundleRequest{
+		TrustDomainId: s.td.String(),
+	})
+	if err != nil {
+		log.WithError(err).Error("Failed to fetch server bundle")
+		return nil, status.Errorf(codes.Internal, "failed to fetch server bundle: %v", err)
+	}
+
+	if dsResp.Bundle == nil {
+		log.Error("Failed to fetch server bundle: not found")
+		return nil, status.Errorf(codes.NotFound, "failed to fetch server bundle: not found")
+	}
+
+	applyBundleMask(req.Bundle, req.InputMask)
+
+	var rootCAs []*common.Certificate
```

Comment:
I expect this validation of incoming bundle proto contents to be useful for Create/Update/Set as well, maybe we should put it into a function we can share?